### PR TITLE
feat(AutoControlledComponent): Default values

### LINKED ISSUE #764: Support defaultProps in AutoControlledComponent
This was fixed for the Accordion but likely exists in other auto controlled components. 

From #763:

The `defaultActiveIndex` is overridden in the Accordion on `componentWillMount`.  This is an edge case, but we need to update the `AutoControlledComponent` to defer to default props when calling `trySetState` in `componentWillMount`.

The real solve here is likely adding `static defaultAutoControlledProps`.  Currently, the `AutoControlledComponent` tells devs to use `trySetState` in `componentWillMount` instead of setting defaults for auto controlled props.  This is to prevent the actual defaults from always winning, however, if we had our own auto controlled defaults then we could apply them correctly.

### PR DESCRIPTION
**Pull request by jeffcarbs**
Fixes #764

I noticed a bug using `AutoControlledComponent` where initial state for non-AutoControlled props was getting clobbered in the `componentWillMount` of the ACC. This is because ACC uses `this.state = ...` instead of `this.setState` to set the state.

Example of the bug:
```js
class MyComponent extends AutoControlledComponent {
  state = { someField: 'default' }

  render () => <div />
}

// Actual
shallow(<MyComponent />).state() === {}

// Expected
shallow(<MyComponent />).state() === { someField: 'default' }
```

In fixing that bug, I came across the #764 issue which I realized could be solved simultaneously. The initial state is set within the constructor which means it's available in the ACC's `componentWillMount`.  I updated the `componentWillMount` to use the initial state if the value of the prop was not defined via props.

For example:
```js
class MyComponent extends AutoControlledComponent {
  state = { someField: 'default' }

  static autoControlledProps = ['someField']

  render () => <div />
}

shallow(<MyComponent />).state() === { someField: 'default' }
shallow(<MyComponent defaultSomeField='propDefault' />).state() === { someField: 'propDefault' }
shallow(<MyComponent someField='prop' />).state() === { someField: 'prop' }
```

This seems pretty intuitive to me. The props are being passed directly to state, so to specify a "global default" outside of props, you set the state yourself in the constructor (or use the es7 property initializer, which is basically just sugar on the same thing).

----

_Edit_ As an aside, I've been using `AutoControlledComponent` as a base for some of my own components via:
```js
import AutoControlledComponent from 'semantic-ui-react/dist/commonjs/lib/AutoControlledComponent'
```

I think it's a really cool abstraction and lets you write super flexible/reusable components that either do all the work (control the state) or let the user do it from higher up.

@levithomason - Thoughts on open-sourcing it as a separate package? Its only dependency is `lodash` (which you could probably even drop) so would probably be pretty simple.

### CODE REVIEW

**Comment by levithomason** [comment]
I think this sounds sane, using the initial state as the "default" AC prop values.  Another way of thinking about this is that it is the "default state", since, the props are copied to state immediately.

I'm gonna update this PR to undo the Accordion hacks as well just to ensure it is working correctly.

File: src/lib/AutoControlledComponent.js [hunk 1]
```diff
@@ -30,27 +30,33 @@ const getDefaultPropName = (prop) => `default${prop[0].toUpperCase() + prop.slic
...
-  const prop = props[propName]
-  const defaultProp = props[defaultPropName]
-
-  const hasProp = prop !== undefined
-  const hasDefaultProp = defaultProp !== undefined
-
-  // defaultProps & props
-  if (includeDefaultProps && !hasProp && hasDefaultProp) return defaultProp
-  if (hasProp) return prop
+export const getAutoControlledStateValue = (propName, props, state, includeDefaults = false) => {
```
Comment (levithomason) [inline src/lib/AutoControlledComponent.js:45]: Updated this function to handle initial state.  Since this is the SSOT when it comes to how we derive AC state I think it makes sense here.  It also prevents an extra loop of state in CWM.

File: test/specs/lib/AutoControlledComponent-test.js [hunk 2]
```diff
@@ -32,6 +32,11 @@ describe('extending AutoControlledComponent', () => {
     TestClass = createTestClass({ autoControlledProps: [], state: {} })
   })
 
+  it('does not throw with a `null` state', () => {
```
Comment (levithomason) [inline test/specs/lib/AutoControlledComponent-test.js:35]: Extra test to catch the edge case I ran into.  If you have `state = undefined` React will initialize state as `null`.  This then throws as we look for keys on `null`.  I added a comment about this and a check for it in `getAutoControlledStateValue`.

**Review by levithomason — COMMENTED** [review]
I updated the Accordion to use the new initial state pattern in the constructor.  I consolidated the "default state" logic into the `getAutoControlledStateValue` function as well.  Finally, I added another test to cover an edge case.

LMK if anything looks out of place.

File: src/lib/AutoControlledComponent.js [hunk 3]
```diff
@@ -30,27 +30,33 @@ const getDefaultPropName = (prop) => `default${prop[0].toUpperCase() + prop.slic
...
+  if (propValue !== undefined) return propValue
+
+  if (includeDefaults) {
+    // defaultProps
+    const defaultProp = props[getDefaultPropName(propName)]
+    if (defaultProp !== undefined) return defaultProp
+
+    // initial state
+    // check for an object as React initializes state as `null` if it is `undefined`
+    const initialState = ({}).toString.call(state) === '[object Object]' && state[propName]
```
Comment (jeffcarbs) [inline src/lib/AutoControlledComponent.js]: I'd maybe do something like:
```js
if (state) {
  const initialState = state[propName]
  if (initialState !== undefined) return initialState
}
```

- `state` will only every be: an object, null, or undefined. I think this is a more straightforward check than the `toString` check.
- `if (initialState) return initialState` wouldn't let `null` through, which should be a valid default value.
- It makes it the same style as the `propValue` and `defaultProp` returns.

**Comment by jeffcarbs** [comment]
Left a comment but otherwise I think this is g2g 👍

**Comment by levithomason** [comment]
I try to avoid extra nesting / branch logic when possible.  Perhaps went too far in that case, updated :)

**Comment by levithomason** [comment]
Released in `semantic-ui-react@0.63.1`.
